Implement node deletions

jmmv opened the pull request:
This PR is a complex because of all the required refactoring of `setattr`, but the implementation of the actual unlinking operations is pretty simple once that is done. Let me know if you want me to send the two commits separately for review.

Also, I've left a couple of TODOs in place because I just can't figure out what the type of the lambda should be. Clippy implies that I can remove the redundant closure, but that's not true.

And... we are almost feature complete! The most significant thing left is the ability to apply reconfigurations.

Thanks

taralx commented on src/nodes/mod.rs:
```diff
@@ -88,57 +88,135 @@ pub struct AttrDelta {
+    let result = try_path(path, |p|
+        unistd::fchownat(None, p, uid, gid, unistd::FchownatFlags::NoFollowSymlink));
+    if result.is_ok() {
+        attr.uid = uid.map_or_else(|| attr.uid, u32::from);
```
Does `map_or` work here? attr.uid is probably Copy, so it shouldn't cause an issue.

taralx commented on src/nodes/mod.rs:
```diff
@@ -88,57 +88,135 @@ pub struct AttrDelta {
+    let atime = atime.unwrap_or_else(|| conv::timespec_to_timeval(attr.atime));
+    let mtime = mtime.unwrap_or_else(|| conv::timespec_to_timeval(attr.mtime));
+    if attr.kind == fuse::FileType::Symlink {
+        // TODO(jmmv): Should use futimensat to avoid following symlinks but this function is
```
Not a fan of this. We should at some point conditionalize and this path should fail if you try to do this to a symlink.

taralx commented:
No new push?

jmmv commented:
I think I got distracted with necessary `nix` changes to make this work well and forgot about getting back to the review. Pushed now but can't remember what status this was in; hopefully nothing embarrassing.

FTR, this PR requires https://github.com/nix-rust/nix/pull/968 in order to get the Linux builds working. And https://github.com/nix-rust/nix/pull/967 will hopefully let me address the issue of updating the times of symlinks.

jmmv commented:
Alright, https://github.com/nix-rust/nix/pull/968 was merged into nix so I have updated `Cargo.toml` accordingly here and fixed the builds. They are now green so this should be ready for another pass. Thanks!

taralx commented on src/lib.rs:
```diff
@@ -12,6 +12,13 @@
+
+// We construct complex structures in multiple places, and allowing for redundant field names
+// increases readability.
 #![cfg_attr(feature = "cargo-clippy", allow(redundant_field_names))]
```
Filed https://github.com/rust-lang-nursery/rust-clippy/issues/3422 for this.

jmmv replied:
Thanks! Seems like a good idea.
